using-docker/snap/snapcraft.yaml

```diff
@@ -0,0 +1,145 @@
+
+grade: stable
+confinement: strict
+
+# The daemons are installed as disabled because the docker{-executables} and
+# interfaces are not auto-connecting. They can be installed as enabled if
+# interface auto-connection is enabled either in the Store UI or in the gadget snap.
+apps:
+  tocker:
+    command: usr/bin/tocker_run.sh
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  ticker:
+    command: usr/bin/tocker_compose.sh
+    daemon: simple
+    install-mode: disable
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      # TODO: genericize getting the proper python path
+      PYTHONPATH: $SNAP/docker-bin/lib/python3.6/site-packages:$PYTHONPATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  nginx-volume:
+    command: usr/bin/tocker_volume.sh
+    daemon: simple
+    install-mode: disable
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  jenkins1:
+    command: usr/bin/tocker_jenkins1.sh
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  jenkins2:
+    command: usr/bin/tocker_jenkins2.sh
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  edgex:
+    command: usr/bin/tocker_edgex.sh
+    daemon: simple
+    install-mode: disable
+    plugs: [docker, docker-executables]
+    environment:
+      PATH: $SNAP/docker-bin/bin:$PATH
+      # TODO: genericize getting the proper python path
+      PYTHONPATH: $SNAP/docker-bin/lib/python3.6/site-packages:$PYTHONPATH
+      DOCKER_CONFIG: $SNAP_COMMON/.docker
+
+  rest:
+    command: usr/bin/tocker_rest.sh
+    plugs: [docker]
+
+parts:
+  copy-src:
+    plugin: dump
+    source: src/
+
+  edgex:
+    plugin: nil
+    override-pull: |
+      _release=levski
+      _compose=docker-compose-no-secty
+      curl -fLo edgex-compose.yml \
+         https://raw.githubusercontent.com/edgexfoundry/edgex-compose/${_release}/${_compose}.yml
+
+      install -Dm644 edgex-compose.yml ${CRAFT_PART_BUILD}/edgex-compose.yml
+    override-build: |
+      # Work around a known Apparmor issue:
+      # https://bugs.launchpad.net/snappy/+bug/1908448
+      sed -i '/security_opt\:/{N;d;}' edgex-compose.yml
+      sed -i '/- no-new-privileges\:true/d' edgex-compose.yml
+
+      install -Dm644 edgex-compose.yml \
+        ${CRAFT_PART_INSTALL}/usr/share/composers/edgex-compose.yml
+
+  seed-image:
+    plugin: nil
+    build-packages: [docker.io]
+    override-build: |
+      # If we already have images in the project directory, assume that is all
+      # of the images we want to include
+      for image in images/*.docker; do
+        [ -e ${CRAFT_PROJECT_DIR}/images/*.docker ] || {
```

Comment:
I think this if-statement should probably be outside of the loop. Also, `test` will freak out if it receives more than one argument (which it will, if there's more than one `.docker` file.

Additionally, there's no logic to install `$image`.

A good way to do this might be to install every docker image from `${CRAFT_PROJECT_DIR}/images` to the appropriate location. Then check if there are any images in the folder, and, if not,
`docker save -o "${CRAFT_INSTALL_PART}/usr/share/images/hello-world.docker" hello-world`

Reply:
> Also, test will freak out if it receives more than one argument (which it will, if there's more than one .docker file.

This is why `[` cannot be outside of the loop and instead we iterate over the images which exist; if there are no objects for the glob to expand to the loop runs once trivially, so we avoid that by using a test; if there aren't actually objects there, *then* fetch some objects. The logic is essentially to avoid the problem of unknown names.

> Additionally, there's no logic to install $image.

Yeah, that's a typo 😅 

> ... Then check if there are any images in the folder ...

This may be better... Though I think it's just reversing the logic order?

Reply:
Hmm.. maybe I'm misunderstanding what's going on here, but won't
`[ -e ${CRAFT_PROJECT_DIR}/images/*.docker ]` match every file ending in `.docker` each iteration of the loop?

We could do something like
```sh
# try installing any existing images
_image_count=0
for _image in images/*.docker; do
  # do all the stuff to install the docker image
  : $(( _image_count += 1 ))
done

# if there are no existing images, install the default
[ "$_image_count -gt 0 ] || {
  # start up docker, pull hello-world, and install
}
```

Reply:
Oh shit yes; THAT (`[`) should be on `$image` >.<
Your suggestion seems fine though.

Reply:
I think for the first one, the existence of `_image` is implied by its enumeration in the for loop, so the statements after the `||` are unreachable.